src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList, GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request, BitConverter.GetBytes(id))].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        [ContractMethod(1_00000000, CallFlags.AllowModifyStates)]
+        private void Incentivize(ApplicationEngine engine)
+        {
+            ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+            BigInteger gas = GAS.BalanceOf(engine.Snapshot, Hash) / nodes.Length;
+            if (gas.IsZero) return;
+            foreach (ECPoint node in nodes)
+            {
+                UInt160 to = Contract.CreateSignatureRedeemScript(node).ToScriptHash();
+                GAS.TransferInternal(engine, Hash, to, gas);
+            }
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData)
+        {
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength)
+                throw new ArgumentException();
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
+            item_id.Value = BitConverter.GetBytes(id);
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Request, item_id.Value), new StorageItem(new OracleRequest
+            {
+                Txid = ((Transaction)engine.ScriptContainer).Hash,
+                Url = url,
+                Filter = filter,
+                CallbackContract = engine.CallingScriptHash,
+                CallbackMethod = callback,
+                UserData = BinarySerializer.Serialize(userData, MaxUserDataLength)
+            }));
+            engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_IdList, GetUrlHash(url)), () => new StorageItem(new IdList())).GetInteroperable<IdList>().Add(id);
```

Comment:
Add a notification here,like `engine.SendNotification(Hash, "Request", new Array() { StackItem.FromInterface(request) });`.
Oracle node need to be triggered by this and query data.

Reply:
They don't need to be triggered by enents.

Reply:
How do they know it's time to start working？ By manual？

Reply:
Every block they call GetRequests().

Reply:
So there are some logic in `Persist (Block block)`. This part of code has nott been added yet?

Reply:
Why there's some logic in `Persist()`? The Oracle nodes subscriber to the Persist event. It is the same logic as contract events.

Reply:
Ok.Got it.

Reply:
> Every block they call GetRequests().

Then oracle node may have to filter duplicate request? I think it may be easier for oracle module by event.

Reply:
I agree with Lu Chuan, when the plugin subscribes to the storage event, as long as it judges the execution status and then obtains the notification of the Oracle contract, it can easily obtain or construct the Request based on this information.

Reply:
What do you mean by "duplicate request"?

Reply:
If you are saying the requests with the same url, you can call `GetRequestsByUrl()`.

https://github.com/neo-project/neo/blob/63b3a759f178a6c30d42844553b85f397c8376ad/src/neo/SmartContract/Native/Oracle/OracleContract.cs#L71-L77